crates/sui/src/upgrade_compatibility.rs

```diff
@@ -101,7 +109,127 @@ pub(crate) enum UpgradeCompatibilityModeError {
     },
 }
 
+/// A list of errors that can occur during upgrade compatibility checks.
+#[derive(Debug, Clone, Default)]
+pub struct UpgradeErrorList {
+    errors: Vec<UpgradeCompatibilityModeError>,
+    source: Option<String>,
+}
+
+impl UpgradeErrorList {
+    fn push(&mut self, err: UpgradeCompatibilityModeError) {
+        self.errors.push(err);
+    }
+
+    /// Only keep the errors that break compatibility with the given [`Compatibility`]
+    fn retain_incompatible(&mut self, compatibility: &Compatibility) {
+        self.errors
+            .retain(|e| e.breaks_compatibility(compatibility));
+    }
+
+    /// Print the errors to the console with the relevant source code.
+    fn print_errors(
+        &mut self,
+        compiled_unit_with_source: &CompiledUnitWithSource,
+    ) -> Result<(), Error> {
+        for err in self.errors.clone() {
+            match err {
+                UpgradeCompatibilityModeError::StructMissing { name, .. } => {
+                    self.print_missing_definition(
+                        "Struct",
+                        name.to_string(),
+                        compiled_unit_with_source,
+                    )?;
+                }
+                UpgradeCompatibilityModeError::EnumMissing { name, .. } => {
+                    self.print_missing_definition(
+                        "Enum",
+                        name.to_string(),
+                        compiled_unit_with_source,
+                    )?;
+                }
+                UpgradeCompatibilityModeError::FunctionMissingPublic { name, .. } => {
+                    self.print_missing_definition(
+                        "Function",
+                        name.to_string(),
+                        compiled_unit_with_source,
+                    )?;
+                }
+                UpgradeCompatibilityModeError::FunctionMissingEntry { name, .. } => {
+                    self.print_missing_definition(
+                        "Function",
+                        name.to_string(),
+                        compiled_unit_with_source,
+                    )?;
```

Comment:
nit: You may have a capitalization issue here, given where the kinds show up in the message:

```suggestion
                    self.print_missing_definition(
                        "struct",
                        name.to_string(),
                        compiled_unit_with_source,
                    )?;
                }
                UpgradeCompatibilityModeError::EnumMissing { name, .. } => {
                    self.print_missing_definition(
                        "enum",
                        name.to_string(),
                        compiled_unit_with_source,
                    )?;
                }
                UpgradeCompatibilityModeError::FunctionMissingPublic { name, .. } => {
                    self.print_missing_definition(
                        "public function",
                        name.to_string(),
                        compiled_unit_with_source,
                    )?;
                }
                UpgradeCompatibilityModeError::FunctionMissingEntry { name, .. } => {
                    self.print_missing_definition(
                        "entry function",
                        name.to_string(),
                        compiled_unit_with_source,
                    )?;
```